File: services/api/app/schemas/candidate_block.py
"""Schemas for Candidate Block Generation (Batch 5C)."""

from __future__ import annotations

from datetime import date, datetime
from pydantic import BaseModel, ConfigDict, Field

DEFAULT_CANDIDATE_ADVISORY_NOTE = (
    "This is a computational candidate for planning evaluation. "
    "It is NOT an engineering, traffic-control, or safety approval."
)


class CandidateBlockResponse(BaseModel):
    """Schema representing a candidate maintenance block option."""

    candidate_id: str
    opportunity_id: str | None = None
    task_ids: list[str]
    departments_involved: list[str]
    section_id: str
    window_id: str
    candidate_start: datetime
    candidate_end: datetime
    required_duration_hrs: float
    window_duration_hrs: float
    source_window_status: str | None = None
    computed_feasibility_status: str  # FEASIBLE | TRAIN_CONFLICT | DURATION_INSUFFICIENT
    train_conflict: bool
    train_conflict_count: int
    freight_data_available: bool
    freight_level: str | None = None
    forecast_freight_trains: int | None = None
    forecast_tonnage: float | None = None
    freight_confidence: float | None = None
    resource_check: str  # VERIFIED | UNVERIFIED | UNAVAILABLE
    resources_available: bool
    resource_ids: list[str]
    priority_score: float
    compatibility_score: float
    candidate_score: float
    reasons: list[str]
    warnings: list[str]
    advisory_note: str = DEFAULT_CANDIDATE_ADVISORY_NOTE

    model_config = ConfigDict(from_attributes=True)


class CandidateBlocksListResponse(BaseModel):
    """Paginated list response for candidate blocks."""

    items: list[CandidateBlockResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

    model_config = ConfigDict(from_attributes=True)

File: services/web/src/lib/types/candidate-block.ts
export interface CandidateBlock {
  candidate_id: string;
  opportunity_id?: string | null;
  task_ids: string[];
  departments_involved: string[];
  section_id: string;
  window_id: string;
  candidate_start: string;
  candidate_end: string;
  required_duration_hrs?: number;
  window_duration_hrs?: number;
  block_duration_hrs?: number;
  computed_feasibility_status?: "FEASIBLE" | "TRAIN_CONFLICT" | "DURATION_INSUFFICIENT" | string;
  feasibility_status?: "FEASIBLE" | "TRAIN_CONFLICT" | "DURATION_INSUFFICIENT" | string;
  train_conflict?: boolean;
  train_conflict_count?: number;
  train_conflicts?: number;
  freight_level?: string | null;
  resource_check?: string;
  priority_score: number;
  compatibility_score?: number;
  candidate_score?: number;
  estimated_impact_score?: number;
  reasons: string[];
  warnings?: string[];
  advisory_note?: string;
  // Legacy aliases
  window_start?: string;
  window_end?: string;
}
